File: impls/julia/step1_read_print.jl
#!/usr/bin/env julia

push!(LOAD_PATH, pwd(), "/usr/share/julia/base")
import readline_mod
import reader
import printer

# READ
function READ(str)
    reader.read_str(str)
end

# EVAL
function EVAL(ast, env)
    ast
end

# PRINT
function PRINT(exp)
    printer.pr_str(exp)
end

# REPL
function REP(str)
    return PRINT(EVAL(READ(str), []))
end

while true
    line = readline_mod.do_readline("user> ")
    if line === nothing break end
    try
        println(REP(line))
    catch e
        if isa(e, ErrorException)
            println("Error: $(e.msg)")
        else
            println("Error: $(string(e))")
        end
        bt = catch_backtrace()
        Base.show_backtrace(STDERR, bt)
        println()
    end
end

File: impls/perl/step1_read_print.pl
#!/usr/bin/perl

use strict;
use warnings;
use File::Basename 'dirname';
use lib dirname(__FILE__);

use English '-no_match_vars';

use Readline qw(mal_readline set_rl_mode);
use Reader   qw(read_str);
use Printer  qw(pr_str);

# read
sub READ {
    my $str = shift;
    return read_str($str);
}

# eval
sub EVAL {
    my ($ast) = @_;
    return $ast;
}

# print
sub PRINT {
    my $exp = shift;
    return pr_str($exp);
}

# repl
sub REP {
    my $str = shift;
    return PRINT( EVAL( READ($str) ) );
}

# Command line arguments
if ( $ARGV[0] eq '--raw' ) {
    set_rl_mode('raw');
    shift @ARGV;
}

while ( defined( my $line = mal_readline('user> ') ) ) {
    eval {
        print REP($line), "\n" or die $ERRNO;
        1;
    } or do {
        my $err = $EVAL_ERROR;
        print 'Error: ', $err or die $ERRNO;
    };
}
